application/src/main/java/org/togetherjava/tjbot/commands/reminder/RemindRoutine.java

```diff
@@ -0,0 +1,142 @@
+        // If guild channel can still be found, send there
+        TextChannel channel = jda.getTextChannelById(channelId);
+        if (channel != null) {
+            return createGuildReminderRoute(jda, authorId, channel);
+        }
+
+        // Otherwise, attempt to DM the user directly
+        return createDmReminderRoute(jda, authorId);
+    }
+
+    private static @NotNull RestAction<ReminderRoute> createGuildReminderRoute(@NotNull JDA jda,
+            long authorId, @NotNull TextChannel channel) {
+        return jda.retrieveUserById(authorId)
+            .onErrorMap(error -> null)
+            .map(author -> ReminderRoute.toPublic(channel, author));
+    }
+
+    private static @NotNull RestAction<ReminderRoute> createDmReminderRoute(@NotNull JDA jda,
+            long authorId) {
+        return jda.openPrivateChannelById(authorId)
+            .map(channel -> ReminderRoute.toPrivate(channel, channel.getUser()));
+    }
+
+    private static void sendReminderViaRoute(@NotNull RestAction<ReminderRoute> routeAction,
+            long id, @NotNull CharSequence content, @NotNull TemporalAccessor createdAt) {
+        Function<ReminderRoute, MessageAction> sendMessage = route -> route.channel
+            .sendMessageEmbeds(createReminderEmbed(content, createdAt, route.target()))
+            .content(route.description());
+
+        Consumer<Throwable> logFailure = failure -> logger.warn(
+                "Failed to send a reminder (id '{}'), skipping it. This can be due to a network issue,"
+                        + " but also happen if the bot disconnected from the target guild and the"
+                        + " user has disabled DMs or has been deleted.",
+                id);
+
+        routeAction.flatMap(sendMessage).queue(doNothing(), logFailure);
+    }
+
+    private static @NotNull MessageEmbed createReminderEmbed(@NotNull CharSequence content,
+            @NotNull TemporalAccessor createdAt, @Nullable User author) {
+        String authorName = author == null ? "Unknown user" : author.getAsTag();
+        String authorIconUrl = author == null ? null : author.getAvatarUrl();
+
+        return new EmbedBuilder().setAuthor(authorName, null, authorIconUrl)
+            .setDescription(content)
+            .setFooter("reminder from")
+            .setTimestamp(createdAt)
+            .setColor(AMBIENT_COLOR)
+            .build();
+    }
+
+    private static <T> @NotNull Consumer<T> doNothing() {
+        return a -> {
+        };
+    }
+
+    private record ReminderRoute(@NotNull MessageChannel channel, @Nullable User target,
+            @Nullable String description) {
+        static ReminderRoute toPublic(@NotNull TextChannel channel, @Nullable User target) {
+            return new ReminderRoute(channel, target,
+                    target == null ? null : target.getAsMention());
+        }
+
+        static ReminderRoute toPrivate(@NotNull PrivateChannel channel, @NotNull User target) {
```

Comment:
In the future this might require some handling since the PrivateChannel#getUser method will become nullable.
This is more of a note for in the future, could add a TODO with a note so we don't forget that?

Reply:
in which version? If we add such a TODO, we need to explicitly link the "when" somehow. otherwise its causing too much confusion when its not coming in the next few weeks

Reply:
JDA 5, no "when" since there's no ETA

Reply:
yeah but now i can at least give it a name. thanks

Reply:
what does it mean when `getUser` on private channel is `null` in JDA 5? maybe i can implement a safe logic directly now already, which would work for both versions.

Reply:
It's nullable, the exact causation I do not remember, something with Discord changing iirc. 
A solution is to retrieve the user first, and use `User#openPrivateChannel`.
But that's an extra API call we might not even need, so whenever you add that is your decision.

Reply:
Okay, thanks for the info. I would be interested in the cause of it being nullable though. If you can open a private channel, the user to that channel should exist lol. Unless they change the meaning of private channel maybe.

Reply:
After doing some quick research, I can explain why, and this shouldn't matter so a comment isn't required.

To elaborate on this:

The MESSAGE_CREATE event gives information of the person that sent the message
The REACTION_ADD event **only** gives the ID of the person that adds the reaction

When you open a private channel, the user and such get retrieved, the info is there.


The issue starts if your bot becomes big, you'll start to use multiple shards. (shard = JDA instance, allows you to run tjhe bot on separate processes / machines)

**All** DM events get send to shard 0, if you open DM's in shard 1, shard 0 **doesn't** have any info about the DM's.
We then get the event, but the event only gives information of the person that triggers it. 

This means the person the DM's are from is unknown.



### Don't worry!
Because our bot is < 3000 (out of my head, could be wrong) guilds this doesn't matter, if our bot was using multiple shards this would be an issue.

So you can ignore this, luckily

Reply:
Makes sense. Thank you for investigating! :)